plugins/watchdog/plugin.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/containerd/log"
+	"github.com/containerd/plugin"
+	"github.com/containerd/plugin/registry"
+	"github.com/coreos/go-systemd/v22/daemon"
+)
+
+type service struct {
+	wb *watchdogNotifier
+}
+
+func init() {
+	registry.Register(&plugin.Registration{
+		Type: plugins.WatchdogPlugin,
+		ID:   "watchdog",
+		Requires: []plugin.Type{
+			plugins.GRPCPlugin,
+		},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			hcp, err := ic.GetByID(plugins.GRPCPlugin, "healthcheck")
+			if err != nil {
+				return nil, err
+			}
+			hc := hcp.(*healthcheck.Service)
+			wn, err := newWatchdogNotifier(hc.IsServing)
+			if err != nil {
+				return nil, fmt.Errorf("cannot init watchdog plugin: %v: %w", err, plugin.ErrSkipPlugin)
+			}
+			// start notifying
+			go wn.Notify(ic.Context)
+			return &service{wn}, nil
+		},
+	})
+}
+
+type watchdogNotifier struct {
+	wc       watchdogClient
+	interval time.Duration
+	healthy  func(ctx context.Context) (bool, error)
+	done     chan struct{}
+}
+
+// option defines optional parameters for initializing the healthChecker
+// structure.
+type option func(*watchdogNotifier)
+
+func withWatchdogClient(wc watchdogClient) option {
+	return func(wn *watchdogNotifier) {
+		wn.wc = wc
+	}
+}
+
+// newWatchdogNotifier creates a new watchogNotifier.
+//
+// If the watchdog is not enabled, the function returns an error.
+func newWatchdogNotifier(healthy func(ctx context.Context) (bool, error), opts ...option) (*watchdogNotifier, error) {
+	wn := &watchdogNotifier{
+		healthy: healthy,
+		wc:      &defaultWatchdogClient{},
+		done:    make(chan struct{}),
+	}
+	for _, o := range opts {
+		o(wn)
+	}
+
+	watchdogVal, err := wn.wc.sdWatchdogEnabled()
+	switch {
+	case err != nil:
+		return nil, fmt.Errorf("cannot determine if watchdog is enabled: %v", err)
+	case watchdogVal == 0:
+		return nil, fmt.Errorf("systemd watchdog not enabled for this PID")
+	}
+	// It is recommended that a daemon sends a keep-alive notification
+	// message to the service manager every half of the time of
+	// WATCHDOG_USEC. See:
+	// https://www.freedesktop.org/software/systemd/man/latest/sd_watchdog_enabled.html
+	wn.interval = watchdogVal / 2
+	return wn, nil
+}
+
+func (wn *watchdogNotifier) Notify(ctx context.Context) {
+	ticker := time.NewTicker(wn.interval)
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case <-ticker.C:
+			isServing, err := wn.healthy(ctx)
+			if err != nil {
+				log.G(ctx).WithError(err).Error("cannot determine if the daemon is healthy")
+				break loop
+			}
+			if !isServing {
+				log.G(ctx).Warn("daemon detected as unhealthy")
+				break loop
+			}
+			ack, err := wn.wc.sdNotify()
+			if err != nil {
+				// Notification not supported. This should never happen.
+				log.G(ctx).WithError(err).Errorf("cannot notify systemd watchdog: %v", err)
+				break loop
+			}
+			if !ack {
+				log.G(ctx).Warn("cannot notify systemd watchdog. retrying")
+				// partition the interval in p times and retry p times
```

Comment:
Is this true? If I'm understanding correctly, the loop below would never continue beyond the first iteration because you `continue` the outer labeled loop. Is this inner loop necessary?

Reply:
No, it is not. 

> Is this inner loop necessary?

Not really. We can just remove it and let the loop retry alone until systemd allows it. Thanks for pointing it out.